### PR TITLE
Fix to avoid error "uninitialized constant ActionController::DataStre…

### PR DESCRIPTION
**Pull request by mbouchard**
…aming::FileBody" when the filename is wrong.

It now returns a 404 instead.

### CODE REVIEW

**Comment by jimryan** [comment]
Could you explain this one to me?  Why isn't the constant defined if the filename is wrong?

**Comment by mbouchard** [comment]
After some testing, I have found that the FileBody class has been moved to another module.

https://github.com/rails/rails/commit/69009f4473637a44ade26d954ef5ddea6ff903f2#diff-9c0a4cd7ac798b3232ea0e418e6f0f6a

That means the gem is not Rails 5 compatible. I have fixed this in the pull request.

The "uninitialized constant" error was not because the filename was wrong (I will provide another fix for this!).

My pull request cannot pass the Travis test because the testing app is not Rails 5+.

**Comment by mbouchard** [comment]
Commit https://github.com/room118solutions/dynamic_paperclip/pull/3/commits/ad6c209cdd08efb202140faaa6aa647987726409 is to fix the following error in Rails 5.1:

lib/ruby/gems/2.4.0/gems/actionpack-5.1.1/lib/action_dispatch/middleware/stack.rb:35:in `build': undefined method `new' for "DynamicPaperclip::AttachmentStyleGenerator":String (NoMethodError)

https://github.com/rails/rails/commit/83b767ce

File: dynamic_paperclip.gemspec [hunk 1]
```diff
@@ -20,12 +20,12 @@ Gem::Specification.new do |s|
   s.add_dependency "paperclip", ">= 3.5.1"
   s.add_dependency 'activesupport'
 
-  # Required for ActionController::DataStreaming::FileBody
-  s.add_dependency "actionpack", ">= 3.2"
+  # Required for ActionDispatch::Response::FileBody
+  s.add_dependency "actionpack", ">= 5.0"
```
Comment (jimryan) [inline dynamic_paperclip.gemspec]: I'd like to continue supporting Rails 4.x, so we should revert this and find a different way to rely upon the correct class

File: lib/dynamic_paperclip/attachment_style_generator.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 require 'rack/request'
-require 'action_controller/metal/data_streaming'
+require 'action_dispatch/http/response.rb'
```
Comment (jimryan) [inline lib/dynamic_paperclip/attachment_style_generator.rb]: It looks like both files exist in both versions, so we could require both, or we could check the Actionpack version and require the one we need.

File: lib/dynamic_paperclip/attachment_style_generator.rb [hunk 3]
```diff
@@ -37,7 +37,7 @@ def call(env)
                 'Content-Transfer-Encoding' => 'binary',
                 'Content-Disposition' => "inline; filename=#{File.basename(attachment.path(style_name))}"
               },
-              ActionController::DataStreaming::FileBody.new(attachment.path(style_name))
+              ActionDispatch::Response::FileBody.new(attachment.path(style_name))
```
Comment (jimryan) [inline lib/dynamic_paperclip/attachment_style_generator.rb]: We should allow for both constants.  We could do something like:
```ruby
file_body = defined?(ActionController::DataStreaming::FileBody) ? ActionController::DataStreaming::FileBody : ActionDispatch::Response::FileBody
file_body.new(attachment.path(style_name))
```

**Comment by mbouchard** [comment]
Sorry for the long delay, I was on holiday. Can you check the changes I made to support both Rails 4 and 5?

**Comment by mbouchard** [comment]
Did you have time to review the changes I've submitted? I developed a new gem that depends on yours and it would be great if those fixes were merged? Thank you!

**Comment by jimryan** [comment]
No I haven't - sorry for the slow review. I'll try to look it over this week. 

Is the test suite passing?

**Comment by mbouchard** [comment]
It seems so.

**Comment by sasselin** [comment]
What about this pull request? I need it in rails 5.1 project.

**Comment by jimryan** [comment]
@mbouchard Sorry for the wait on this, I completely forgot about it.  Thank you so much for your contribution.

**Pull request merged**